src/swapd/runtime.rs

```diff
@@ -210,16 +210,71 @@ pub struct Runtime {
 }
 
 // FIXME Something more meaningful than ServiceId to index
-#[derive(Default)]
-pub struct PendingRequests(HashMap<ServiceId, Vec<PendingRequest>>);
+type PendingRequests = HashMap<ServiceId, Vec<PendingRequest>>;
 
-impl From<HashMap<ServiceId, Vec<PendingRequest>>> for PendingRequests {
-    fn from(m: HashMap<ServiceId, Vec<PendingRequest>>) -> Self {
-        PendingRequests(m)
+impl PendingRequestsT for PendingRequests {
+    fn defer_request(&mut self, key: ServiceId, pending_req: PendingRequest) {
+        let pending_reqs = self.entry(key).or_insert(vec![]);
+        pending_reqs.push(pending_req);
+    }
+    fn continue_deferred_requests(
+        runtime: &mut Runtime,
+        endpoints: &mut Endpoints,
+        key: ServiceId,
+        predicate: fn(&PendingRequest) -> bool,
+    ) -> bool {
+        let success = if let Some(pending_reqs) = runtime.pending_requests.remove(&key) {
+            let len0 = pending_reqs.len();
+            let remaining_pending_reqs: Vec<_> = pending_reqs
+                .into_iter()
+                .filter_map(|r| {
+                    if predicate(&r) {
+                        if let Ok(_) = match &r.bus_id {
+                            ServiceBus::Ctl if &r.dest == &runtime.identity => runtime
+                                .handle_rpc_ctl(endpoints, r.source.clone(), r.request.clone()),
+                            ServiceBus::Msg if &r.dest == &runtime.identity => runtime
+                                .handle_rpc_msg(endpoints, r.source.clone(), r.request.clone()),
+                            _ => endpoints
+                                .send_to(
+                                    r.bus_id.clone(),
+                                    r.source.clone(),
+                                    r.dest.clone(),
+                                    r.request.clone(),
+                                )
+                                .map_err(Into::into),
+                        } {
+                            None
+                        } else {
+                            Some(r)
+                        }
+                    } else {
+                        Some(r)
+                    }
+                })
+                .collect();
+            let len1 = remaining_pending_reqs.len();
+            runtime.pending_requests.insert(key, remaining_pending_reqs);
+            if len0 - len1 > 1 {
+                error!("consumed more than one request with this predicate")
+            }
+            len0 > len1
+        } else {
+            error!("no request consumed with this predicate");
+            false
+        };
+        success
     }
 }
 
-impl PendingRequests {}
+trait PendingRequestsT {
```

Comment:
What is the purpose of this trait? Is this interface reused somewhere else?

Reply:
Go it!

Reply:
Common pattern to extend features from foreign types, the most often I saw it as `WriteExt: std::io::Write` to extend writer capabilities (e.g. in encoding/decoding rust-bitcoin).